src/classes/CW_AudienceCSVUploadController.cls

```diff
@@ -0,0 +1,313 @@
+			return null;
+		}
+       	if(filename==null || filename=='')
+       	{
+         	Apexpages.addMessage(new Apexpages.Message(ApexPages.Severity.Error,System.Label.CW_Error_FileNameMissing));
+			csvFile=null;
+			filename=null;
+	        return null;
+		}
+       
+		CSVBatchReader batchReader;
+		//Validate file context 
+		try
+		{
+			String fileHeader=System.EncodingUtil.convertToHex(csvFile).Substring(0,12);
+         	
+         	//check if the file is an zip archive(propably new excell format or Open Office) or is an excel file 
+			if(fileHeader.startsWithIgnoreCase('504b03') || fileHeader.startsWithIgnoreCase('d0cf11e0a1'))
+			{
+           		Apexpages.addMessage(new Apexpages.Message(ApexPages.Severity.Error,System.Label.CW_Error_BinaryFile));
+	            csvFile=null;
+	            filename=null;
+	            return null;
+         	}
+         	
+         	String fileContent=csvFile.toString();
+         	List<String> lines = new List<String>();
+            
+         	for(Integer sp: notAllowedChr)
+         	{
+           		List<Integer> oneChar=new list<Integer>();
+           		oneChar.Add(sp);
+           		String spChar=String.fromCharArray(oneChar);
+           		if(fileContent.IndexOf(spChar)>0)
+           		{
+             		Apexpages.addMessage(new Apexpages.Message(ApexPages.Severity.Error,System.Label.CW_Error_InvalidChar));
+             		return null;
+           		}
+         	}       
+         
+           	batchReader = new CSVBatchReader(fileContent,campaignId, selectedSchId);
+
+       	}catch(Exception Ex){
+         	ApexPages.addMessages(ex);
+         	csvFile=null;
+         	filename=null;
+         	return null;
+       	}
+       	
+       	csvFile = null;
+       	batchJobId = Database.executeBatch(batchReader,BATCH_SIZE);
+       	return loadBatchJobProgressPage();
+    }
+    
+    public Pagereference loadBatchJobProgressPage()
+    {
+        Pagereference progressPage = Page.CW_BatchJobProgressPage;
+        progressPage.setRedirect(true);
+        progressPage.getParameters().put(ConstantUtil.PAGEPARAM_BATCHJOBID,batchJobId);
+        progressPage.getParameters().put(ConstantUtil.PAGEPARAM_CAMPAIGNID,campaignId);
+        
+        return progressPage;
+    }
+ 
+    
+    /*** static methods for processing records ***/
+    
+    // CSV only contains group name
+    public static void processGroups(List<List<string>> csvLines, Id campaignId,Id scheduleId)
+    {
+        List<CW_Audience_Group__c> aGroups = new List<CW_Audience_Group__c>();
+        List<Account> accounts = new List<Account>();
+
+        Set<String> groupNames = new Set<String>();
+        Set<String> accountNames = new Set<String>();
+        Set<String> agsCompKey = new Set<String>();
+		//get record type id from person accont
+        Id personAccountRTId = [select id from RecordType where SobjectType = 'Account' AND IsPersonType = true limit 1].id;
```

Comment:
Problem: Unbound SOQL query. Reason: a SOQL query can return up to 50,000 records. If this number is exceeded a runtime exception is thrown. Suggestion: cap the max number of records returned adding a LIMIT clause to the query.